pootle/apps/pootle_store/diff.py

```diff
@@ -49,53 +52,117 @@ def target(self):
+    db_unit_class = DBUnit
+
+    unit_fields = (
+        "unitid", "state", "id", "index", "revision",
+        "source_f", "target_f", "developer_comment",
+        "translator_comment", "locations", "context")
+
+    def __init__(self, target_store, source_store):
+        self.target_store = target_store
+        self.source_store = source_store
+
+    def get_db_units(self, unit_qs):
+        diff_units = OrderedDict()
+        units = unit_qs.values(*self.unit_fields).order_by("index")
+        for unit in units:
+            diff_units[unit["unitid"]] = unit
+        return diff_units
+
+    def get_file_unit(self, unit):
+        state = UNTRANSLATED
+        if unit.isobsolete():
+            state = OBSOLETE
+        elif unit.istranslated():
+            state = TRANSLATED
+        elif unit.isfuzzy():
+            state = FUZZY
+        return {
+            "unitid": unit.getid(),
+            "context": unit.getcontext(),
+            "locations": unit.getlocations(),
+            "source": unit.source,
+            "target": unit.target,
+            "state": state,
+            "developer_comment": unit.getnotes(origin="developer"),
+            "translator_comment": unit.getnotes(origin="translator")}
+
+    def get_file_units(self, units):
+        diff_units = OrderedDict()
+        for unit in units:
+            if unit.isheader():
+                continue
+            diff_units[unit.getid()] = self.get_file_unit(unit)
+        return diff_units
+
+    @cached_property
+    def target_units(self):
+        return self.get_db_units(self.target_store.unit_set)
+
+    @cached_property
+    def source_units(self):
+        if isinstance(self.source_store, models.Model):
+            return self.get_db_units(self.source_store.unit_set)
+        return self.get_file_units(self.source_store.units)
+
+    @property
+    def target_unit_class(self):
+        return self.db_unit_class
+
+    @property
+    def source_unit_class(self):
+        if isinstance(self.source_store, models.Model):
+            return self.db_unit_class
+        return self.file_unit_class
+
+
 class StoreDiff(object):
+    """Compares 2 DBStores"""
+
+    def __init__(self, target_store, source_store, source_revision):
+        self.target_store = target_store
+        self.source_store = source_store
+        self.source_revision = source_revision
+
+    @property
+    def diff_class(self):
+        diffs = format_diffs.gather()
+        differ = diffs.get(
+            self.target_store.filetype.name)
+        if differ:
+            return differ
+        return diffs["default"]
 
-    def __init__(self, db_store, file_store, file_revision):
-        self.db_store = db_store
-        self.file_store = file_store
-        self.db_revision = db_store.get_max_unit_revision()
-        self.file_revision = file_revision
+    @cached_property
+    def target_revision(self):
+        return self.target_store.get_max_unit_revision()
```

Comment:
IIRC we should get db_revision before we start update process to avoid any of race conditions.
I'm not clear when we call this method first time in the current implementation.